File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    s = input()
    stack = []
    top = -1
    result = ''
    if s == '.':
        break
    for i in s:
        if i == '(' or i == '[':
            stack.append(i)
            top += 1
        elif i == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
                top -= 1
            else:
                stack.append(i)
                break
        elif i == ']':
            if len(stack) != 0 and stack[-1] == '[':
                stack.pop()
                top -= 1
            else:
                stack.append(i)
                break
    if len(stack) == 0:
        result = 'yes'
    else:
        result = 'no'
    print(result)

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static boolean check(String word){
        Stack<Character> stack = new Stack<>();
        for(char c : word.toCharArray()) {
            if(c == '(' || c == '['){
                stack.push(c);
            } else if (c == ')') {
                if(stack.isEmpty() || stack.pop() != '('){
                    return false;
                }
            } else if (c == ']'){
                if(stack.isEmpty() || stack.pop() != '['){
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while(true){
            String word = br.readLine();
            if (word.equals(".")){
                break;
            }
            if(check(word)){
                System.out.println("yes");
            } else{
                System.out.println("no");
            }


        }
    }
}
